### PR TITLE
feat: add access to module functionalities from the extension

### PR DESCRIPTION
**Pull request by bartoszherba**
closing [IN-3963](https://vsf.atlassian.net/browse/IN-3963)
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSDoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

[IN-3963]: https://vsf.atlassian.net/browse/IN-3963?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/content/4.sdk/3.advanced/extending-module.md [hunk 1]
```diff
@@ -237,6 +237,8 @@ const sapccPaymentConfig = sdk.sapcc.utils.buildConfig(baseConfig);
 Like the built-in SDK methods, methods in `extend` are impacted by your interceptors.
 :::
 
+The very common use case and also very basic example of using `extend` is to create a new method that is not covered by the module. In this example the `extend` attribute is configured with an object that contains a method `getProductBySku` that returns a promise.
```
Comment (lsliwaradioluz) [inline docs/content/4.sdk/3.advanced/extending-module.md:240]:
```suggestion
The very common use case and also very basic example of using `extend` is to create a new method that is not covered by the module. In this example the `extend` attribute is configured with an object containing a method `getProductBySku` that returns a promise.
```

File: docs/content/4.sdk/3.advanced/extending-module.md [hunk 2]
```diff
@@ -260,6 +262,50 @@ import { sdk } from './sdk';
...
+          );
+        },
+      },
+    };
+  }),
+};
+```
+In such way you can easily extend the module with new methods and use the original methods. Also, it is up to the developer to decide if the `context` object is available and what properties it contains.
+
+:::tip As a rule of thumb, it's recommended to add options and client to the context object as it allows for easy implementation of custom methods.
```
Comment (lsliwaradioluz) [inline docs/content/4.sdk/3.advanced/extending-module.md:302]:
```suggestion
:::tip
As a rule of thumb, it's recommended to add options and client to the context object as it allows for easy implementation of custom methods.
```

File: docs/content/4.sdk/3.advanced/extending-module.md [hunk 3]
```diff
@@ -260,6 +262,50 @@ import { sdk } from './sdk';
...
+:::
+
+:::warning the `context` object is optional and might not be available in all modules. Please, check the module's documentation to see if it's available and what properties it contains. You can also check type-hinting in your IDE to see what properties are available.
```
Comment (lsliwaradioluz) [inline docs/content/4.sdk/3.advanced/extending-module.md:305]:
```suggestion
:::warning
The `context` object is optional and might not be available in all modules. Please, check the module's documentation to see if it's available and what properties it contains. You can also check type-hinting in your IDE to see what properties are available.
```

File: packages/sdk/src/bootstrap.ts [hunk 4]
```diff
@@ -65,7 +65,7 @@ export const initSDK = <T extends SDKConfig>(sdkConfig: T): SDKApi<T> => {
         const methodFromExtend = Reflect.get(extend, propKey, receiver);
         const methodFromTarget = Reflect.get(target, propKey, receiver);
 
-        const method = methodFromExtend ?? methodFromTarget;
+        const method = methodFromTarget ?? methodFromExtend;
```
Comment (WojtekTheWebDev) [inline packages/sdk/src/bootstrap.ts:68]: Worth to add a comment here, me and @bartoszherba thinks this was unintended before, `extend` should work only for non-existing methods as we have `override`, so this fix is not related with the core of the feature but worth to add.

**Review by lsliwaradioluz — APPROVED** [review]

File: docs/content/4.sdk/3.advanced/extending-module.md [hunk 5]
```diff
@@ -260,6 +262,50 @@ import { sdk } from './sdk';
...
+
+```ts
+const sdkConfig = {
+  sapcc: buildModule(
+    // sapccModule is the module that we want to extend
+    sapccModule, 
+    // module options
+    { url: "some-url" }, 
+    // factory function that receives the options and the parent object as arguments
+    (options, parent) => {
```
Comment (WojtekTheWebDev) [inline docs/content/4.sdk/3.advanced/extending-module.md:281]: I think that `parent` is a confusing name, maybe we should document it as a `module` or `{ methods, context }`?

File: docs/content/4.sdk/3.advanced/extending-module.md (hunk 2 above)
Comment (WojtekTheWebDev) [inline docs/content/4.sdk/3.advanced/extending-module.md:302]:
```suggestion
:::tip 
As a rule of thumb, it's recommended to add options and client to the context object as it allows for easy implementation of custom methods.
```

File: docs/content/4.sdk/3.advanced/extending-module.md (hunk 3 above)
Comment (WojtekTheWebDev) [inline docs/content/4.sdk/3.advanced/extending-module.md:305]:
```suggestion
:::warning 
The `context` object is optional and might not be available in all modules. Please, check the module's documentation to see if it's available and what properties it contains. You can also check type-hinting in your IDE to see what properties are available.
```

File: docs/content/4.sdk/3.advanced/extending-module.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jagoral) [reply docs/content/4.sdk/3.advanced/extending-module.md:281]: great! I thought that they will be typed as any :D good job

Comment (jagoral) [inline docs/content/4.sdk/3.advanced/extending-module.md:281]: wdyt about adding an example how to define the factory function outside the `buildModule`? When customer will have a lot of modules, he may want to move the factory to a variable and then pass this variable to the buildModule

**Review by jagoral — APPROVED** [review]
looks good!

File: packages/sdk/src/bootstrap.ts (hunk 4 above)
Comment (bartoszherba) [reply packages/sdk/src/bootstrap.ts:68]: Yes, forgot to comment this line.

File: docs/content/4.sdk/3.advanced/extending-module.md (hunk 5 above)
Comment (bartoszherba) [reply docs/content/4.sdk/3.advanced/extending-module.md:281]: @lsliwaradioluz @Fifciu WDYT? I think that `module` would be ok. I wanted to smuggle the hierarchy in that name.

**Pull request merged**